keys.js

```diff
@@ -76,5 +77,37 @@ define(function(require) {
 		}
 	}
 
+	/**
+	 * Resolve all key-value pairs in the supplied object and return a promise
+	 * that will always fulfill with the outcome states of all input promises.
+	 * @param {object} object whose key-value pairs will be settled
+	 * @returns {Promise} promise for an object with the mapped and fully
+	 *  settled key-value pairs
+	 */
+	 function settle(object) {
+			var p = Promise._defer();
+			var resolver = Promise._handler(p);
+
+			var results = {};
+			var keys = Object.keys(object);
+			var promises = keys.map(function(k) { return object[k]; });
+			var pending = keys.length;
+
+			when.settle(promises).then(function(d) {
+				for(var i=0; i<keys.length; i++) {
+					results[keys[i]] = d[i];
+					if(--pending === 0) {
+						resolver.resolve(results);
+					}
+				}
+			});
```

Comment:
I think this can be simplified a bit by getting rid of the `pending` count, and hoisting out the body of the closure.  I think it's also possible to check `keys.length` early and bail out if it's zero.  Shooting from the hip, here's a quick (and untested!) take on it:

```js
function settle(object) {
   var keys = Object.keys(object);
   var results = {};

   if(keys.length === 0) {
     return toPromise(results);
   }

   var p = Promise._defer();
   var resolver = Promise._handler(p);
   var promises = keys.map(function(k) { return results[k]; });

   when.settle(promises).then(function(states) {
     populateResults(keys, states, results, resolver);
   });

   return p;
}

function populateResults(keys, states, results, resolver) {
  for(var i=0; i<keys.length; i++) {
    results[keys[i]] = states[i];
  }
  resolver.resolve(results);  
}
```

I probably didn't get that exactly right, but wanted to see what you think.

It does several passes over the keys (4, I think: `keys.map`, `when.settle` does 2, then the loop in `populateResults`).  It'd be great to find a way to do it in fewer passes, but it's not obvious to me yet how to do it in a compact way.  Any ideas?